src/Processors/Transforms/AggregatingTransform.h

```diff
@@ -70,6 +71,46 @@ struct ManyAggregatedData
         for (auto & mut : mutexes)
             mut = std::make_unique<std::mutex>();
     }
+
+    ~ManyAggregatedData()
+    {
+        try
+        {
+            if (variants.size() <= 1)
+                return;
+
+            // Aggregation states destruction may be very time-consuming.
+            // In the case of a query with LIMIT, most states won't be destroyed during conversion to blocks.
+            // Without the following code, they would be destroyed in the destructor of AggregatedDataVariants in the current thread (i.e. sequentially).
+            const auto pool = std::make_unique<ThreadPool>(variants.size());
+
+            for (auto && variant : variants)
+            {
+                if (variant->size() < 100'000) // some seemingly reasonable constant
+                    continue;
+
+                // It doesn't make sense to spawn a thread if the variant is not going to actually destroy anything.
+                if (variant->aggregator)
+                {
+                    // variant is moved here and will be destroyed in the destructor of the lambda function.
+                    pool->trySchedule(
+                        [variant = std::move(variant), thread_group = CurrentThread::getGroup()]()
+                        {
+                            if (thread_group)
+                                CurrentThread::attachToIfDetached(thread_group);
+
+                            setThreadName("AggregDestruct");
+                        });
+                }
+            }
+
+            pool->wait();
```

Comment:
I often observed that only one thread for a big variant was running for long time while other threads finished immediately or not created at all. The result is that only one thread are do destruction and the current caller thread wait for the only one worker thread. 

For example, the below screenshot shows the scenario that main thread(1431141) is waiting while only one worker thread(1432649) is working. It does not gain performance improvement.

![屏幕截图 2022-10-26 201957](https://user-images.githubusercontent.com/1518453/198025757-c26c1e94-7afe-4979-98a2-b40da97dd1ed.png)

Can I put the destruction to a thread and not wait for it?

Reply:
first of all we need to reproduce it. could you pls share table, query and logs for that query in #42661?

Reply:
@nickitat  After my investigation I found it is because my customization. I suppress the “output to column” step during aggregation, that skipped destruction of "places“. Without my modification, the destruction occurs in parallel and very efficient. Good.